File: mouse_windows/src/mouse_windows.c
#include "mouse_windows.h"

void MouseMove(float x, float y) {
  POINT p;
  if (GetCursorPos(&p)) {
    SetCursorPos(p.x + (int)x, p.y + (int)y);
  }
}

void MouseMoveTo(float x, float y) {
  SetCursorPos((int)x, (int)y);
}

void DoubleClick(void) {
  INPUT inputs[2] = {0};

  // First click
  inputs[0].type = INPUT_MOUSE;
  inputs[0].mi.dwFlags = MOUSEEVENTF_LEFTDOWN;

  inputs[1].type = INPUT_MOUSE;
  inputs[1].mi.dwFlags = MOUSEEVENTF_LEFTUP;

  SendInput(ARRAYSIZE(inputs), inputs, sizeof(INPUT));

  // Small delay
  Sleep(100);

  // Second click
  inputs[0].mi.dwFlags = MOUSEEVENTF_LEFTDOWN;
  inputs[1].mi.dwFlags = MOUSEEVENTF_LEFTUP;

  SendInput(ARRAYSIZE(inputs), inputs, sizeof(INPUT));
}

void MouseScroll(int x, int y, int amount) {
  INPUT input = {0};
  input.type = INPUT_MOUSE;

  if (y != 0) {
    input.mi.dwFlags = MOUSEEVENTF_WHEEL;
    input.mi.mouseData = amount * y;
  } else if (x != 0) {
    input.mi.dwFlags = MOUSEEVENTF_HWHEEL;
    input.mi.mouseData = amount * x;
  }

  SendInput(1, &input, sizeof(INPUT));
}

void MouseClick(MouseButton button) {
  INPUT inputs[2] = {0};

  inputs[0].type = INPUT_MOUSE;
  inputs[1].type = INPUT_MOUSE;

  switch (button) {
    case kMouseButtonLeft:
      inputs[0].mi.dwFlags = MOUSEEVENTF_LEFTDOWN;
      inputs[1].mi.dwFlags = MOUSEEVENTF_LEFTUP;
      break;
    case kMouseButtonRight:
      inputs[0].mi.dwFlags = MOUSEEVENTF_RIGHTDOWN;
      inputs[1].mi.dwFlags = MOUSEEVENTF_RIGHTUP;
      break;
    case kMouseButtonMiddle:
      inputs[0].mi.dwFlags = MOUSEEVENTF_MIDDLEDOWN;
      inputs[1].mi.dwFlags = MOUSEEVENTF_MIDDLEUP;
      break;
  }

  SendInput(ARRAYSIZE(inputs), inputs, sizeof(INPUT));
}

void MouseHoldLeftButton(void) {
  INPUT input = {0};
  input.type = INPUT_MOUSE;
  input.mi.dwFlags = MOUSEEVENTF_LEFTDOWN;

  SendInput(1, &input, sizeof(INPUT));
}

void MouseReleaseLeftButton(void) {
  INPUT input = {0};
  input.type = INPUT_MOUSE;
  input.mi.dwFlags = MOUSEEVENTF_LEFTUP;

  SendInput(1, &input, sizeof(INPUT));
}

File: mouse_windows/lib/mouse_windows.dart

import 'dart:ffi';
import 'mouse_windows_bindings_generated.dart';
import 'package:mouse_platform_interface/mouse_platform_interface.dart' as i;


class MouseWindows extends i.MousePlatform {
  static void registerWith() {
    i.MousePlatform.instance = MouseWindows();
  }

  @override
  void move(double x, double y) => _bindings.MouseMove(x, y);

  @override
  void moveTo(double x, double y) => _bindings.MouseMoveTo(x, y);

  @override
  void holdLeftButton() => _bindings.MouseHoldLeftButton();

  @override
  void releaseLeftButton() => _bindings.MouseReleaseLeftButton();

  @override
  void click(i.MouseButton button) => _bindings.MouseClick(button.value);

  @override
  void doubleClick() => _bindings.DoubleClick();

  @override
  void scroll(int x, int y, int amount) => _bindings.MouseScroll(x, y, amount);
}


const String _libName = 'mouse_windows';

/// The dynamic library in which the symbols for [MouseWindowsBindings] can be found.
final DynamicLibrary _dylib = 
     DynamicLibrary.open('$_libName.dll');
 
/// The bindings to the native functions in [_dylib].
final MouseWindowsBindings _bindings = MouseWindowsBindings(_dylib);

